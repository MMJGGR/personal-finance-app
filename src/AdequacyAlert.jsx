import React, { useMemo } from 'react'
import { useFinance } from './FinanceContext'
import { computeFundingGaps } from './engines/adequacy'
import { formatCurrency } from './utils/formatters'

export default function AdequacyAlert() {
  const { cumulativePV, startYear, settings } = useFinance()
  const gaps = useMemo(() => computeFundingGaps(cumulativePV), [cumulativePV])
  const rows = useMemo(
    () =>
      gaps
        .map((gap, i) => (gap > 0 ? { year: startYear + i, gap } : null))
        .filter(Boolean),
    [gaps, startYear]
  )
  if (rows.length === 0) return null
  return (
    <div
      id="adequacy-alert"
      className="bg-red-50 border border-red-300 p-4 rounded-lg"
    >
      <h3 className="text-red-700 font-semibold mb-2">Adequacy Alert</h3>
      <table className="w-full text-sm">
        <thead>
          <tr className="text-left">
            <th className="py-1 pr-2">Year</th>
            <th className="py-1">Funding Gap</th>
          </tr>
        </thead>
        <tbody>
          {rows.map(r => (
            <tr key={r.year} className="border-t">
              <td className="py-1 pr-2">{r.year}</td>
              <td className="py-1 text-right">
                {formatCurrency(r.gap, settings.locale, settings.currency)}
              </td>
            </tr>
          ))}
        </tbody>
      </table>
    </div>
  )
}
