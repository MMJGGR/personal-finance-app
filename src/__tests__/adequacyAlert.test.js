/* global test, expect, beforeAll, afterEach */
import React from 'react'
import { render, screen } from '@testing-library/react'
import { FinanceProvider, useFinance } from '../FinanceContext'
import AdequacyAlert from '../AdequacyAlert'

beforeAll(() => {
  global.ResizeObserver = class { observe() {} unobserve() {} disconnect() {} }
})

afterEach(() => {
  localStorage.clear()
})

test('renders nothing when no gaps', () => {
  const { container } = render(
    <FinanceProvider>
      <AdequacyAlert />
    </FinanceProvider>
  )
  expect(container.firstChild).toBeNull()
})

test('shows funding gaps table', async () => {
  localStorage.setItem(
    'incomeSources',
    JSON.stringify([{ name: 'Job', amount: 1000, frequency: 1, growth: 0, taxRate: 0 }])
  )
  localStorage.setItem(
    'expensesList',
    JSON.stringify([{ name: 'Big', amount: 1500, paymentsPerYear: 1, growth: 0, priority: 1 }])
  )
  function Wrapper({ children }) {
    const { setYears } = useFinance()
    React.useEffect(() => { setYears(2) }, [setYears])
    return children
  }
  const { container } = render(
    <FinanceProvider>
      <Wrapper>
        <AdequacyAlert />
      </Wrapper>
    </FinanceProvider>
  )
  await screen.findByText('Adequacy Alert')
  expect(screen.getAllByRole('row')).toHaveLength(3)
  expect(container.firstChild).toMatchSnapshot()
})
