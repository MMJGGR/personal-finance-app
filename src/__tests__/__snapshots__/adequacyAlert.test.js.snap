// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shows funding gaps table 1`] = `
<div
  class="bg-red-50 border border-red-300 p-4 rounded-lg"
  id="adequacy-alert"
>
  <h3
    class="text-red-700 font-semibold mb-2"
  >
    Adequacy Alert
  </h3>
  <table
    class="w-full text-sm"
  >
    <thead>
      <tr
        class="text-left"
      >
        <th
          class="py-1 pr-2"
        >
          Year
        </th>
        <th
          class="py-1"
        >
          Funding Gap
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        class="border-t"
      >
        <td
          class="py-1 pr-2"
        >
          NaN
        </td>
        <td
          class="py-1 text-right"
        >
          Ksh 476.19
        </td>
      </tr>
      <tr
        class="border-t"
      >
        <td
          class="py-1 pr-2"
        >
          NaN
        </td>
        <td
          class="py-1 text-right"
        >
          Ksh 929.71
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
